subsys/bluetooth/services/ras/rreq/ras_rreq.c

```diff
@@ -859,3 +860,82 @@ int bt_ras_rreq_cp_get_ranging_data(struct bt_conn *conn, struct net_buf_simple
 
 	return 0;
 }
+
+void bt_ras_rreq_rd_subevent_data_parse(struct net_buf_simple *peer_ranging_data_buf,
+					struct net_buf_simple *local_step_data_buf,
+					bt_ras_rreq_subevent_header_cb_t subevent_header_cb,
+					bt_ras_rreq_step_data_cb_t step_data_cb, void *user_data)
+{
+	if (!peer_ranging_data_buf || !local_step_data_buf) {
+		LOG_ERR("Tried to parse empty step data.");
+		return;
+	}
+
+	/* Remove ranging data header. */
+	net_buf_simple_pull_mem(peer_ranging_data_buf, sizeof(struct ras_ranging_header));
+
+	while (peer_ranging_data_buf->len >= sizeof(struct ras_subevent_header)) {
+		struct ras_subevent_header *peer_subevent_header_data = net_buf_simple_pull_mem(
+			peer_ranging_data_buf, sizeof(struct ras_subevent_header));
+
+		if (subevent_header_cb &&
+		    !subevent_header_cb(peer_subevent_header_data, user_data)) {
+			return;
+		}
+
+		if (peer_subevent_header_data->num_steps_reported == 0 ||
+		    peer_ranging_data_buf->len == 0) {
+			return;
+		}
+
+		for (uint8_t i = 0; i < peer_subevent_header_data->num_steps_reported; i++) {
+			struct bt_le_cs_subevent_step local_step;
+			struct ras_rd_cs_subevent_step peer_step;
+
+			if (local_step_data_buf->len < sizeof(struct bt_le_cs_subevent_step) ||
+			    peer_ranging_data_buf->len < sizeof(struct ras_rd_cs_subevent_step)) {
+				LOG_WRN("Step data appears malformed.");
+			}
+
+			local_step.mode = net_buf_simple_pull_u8(local_step_data_buf);
+			local_step.channel = net_buf_simple_pull_u8(local_step_data_buf);
+			local_step.data_len = net_buf_simple_pull_u8(local_step_data_buf);
+
+			peer_step.mode = net_buf_simple_pull_u8(peer_ranging_data_buf);
+
+			if (peer_step.mode != local_step.mode) {
+				LOG_WRN("Mismatch of local and peer step mode %d != %d",
+					peer_step.mode, local_step.mode);
+				return;
+			}
+
+			if (local_step.data_len == 0) {
+				LOG_WRN("Encountered zero-length step data.");
+				return;
+			}
+
+			peer_step.data = peer_ranging_data_buf->data;
+			local_step.data = local_step_data_buf->data;
+
+			if (local_step.data_len > local_step_data_buf->len ||
+			    peer_ranging_data_buf->len < local_step.data_len) {
+				LOG_WRN("Step data appears malformed.");
+				return;
+			}
+
+			uint16_t peer_step_data_length;
```

Comment:
I wonder if it's possible to infer `peer_step_data_length` automatically from `local_step.mode` and/or `local_step.data_len` so that the user doesn't need to calculate it themselves. Might also need to add N_AP as a parameter to `bt_ras_rreq_rd_subevent_data_parse()`

Right now it seems relatively easy for users to screw up

Reply:
I'll give it a go

Reply:
Updated